background-processor/src/lib.rs

```diff
@@ -0,0 +1,345 @@
+const CHAN_FRESHNESS_TIMER: u64 = 60;
+#[cfg(test)]
+const CHAN_FRESHNESS_TIMER: u64 = 1;
+
+impl BackgroundProcessor {
+	/// Start the background thread that takes care of responsibilities (enumerated in the top-level
+	/// documentation). Marked as `must_use` because otherwise the result is dropped immediately,
+	/// resulting in the thread being terminated.
+	/// Important note: this thread will panic if invoking `persist_manager` results in an error (and
+	/// `start()` will need to be called again to restart the `BackgroundProcessor`).
+	/// There are 3 main options for handling this panic:
+	/// * wait on [`thread_handle`]'s `join()`, handle the error
+	/// * [configure] to abort on panic
+	/// * write a custom `persist_manager` to handle the error so it never gets returned to
+	///   `BackgroundProcessor`.
+	///
+	/// `persist_manager` is responsible for writing out the `ChannelManager` to disk, and/or uploading
+	/// to one or more backup services. See [`ChannelManager::write`] for writing out a `ChannelManager`.
+	/// See [`FilesystemPersister::persist_manager`] for Rust-Lightning's provided implementation.
+	///
+	/// [`thread_handle`]: struct.BackgroundProcessor.html#structfield.thread_handle
+	/// [configure]: https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/aborting-on-panic.html
+	/// [`ChannelManager::write`]: ../lightning/ln/channelmanager/struct.ChannelManager.html#method.write
+	/// [`FilesystemPersister::persist_manager`]: ../lightning_persister/struct.FilesystemPersister.html#impl
+	#[must_use]
+	pub fn start<PM, ChanSigner, M, T, K, F, L>(persist_manager: PM, manager: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>, logger: Arc<L>) -> Self
+	where ChanSigner: 'static + ChannelKeys + Writeable,
+        M: 'static + chain::Watch<Keys=ChanSigner>,
+        T: 'static + BroadcasterInterface,
+        K: 'static + KeysInterface<ChanKeySigner=ChanSigner>,
+        F: 'static + FeeEstimator,
+        L: 'static + Logger,
+        PM: 'static + Send + Fn(Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Result<(), std::io::Error>,
+	{
+		let stop_thread = Arc::new(AtomicBool::new(false));
+		let stop_thread_clone = stop_thread.clone();
+		let handle = thread::spawn(move || {
+			let mut current_time = Instant::now();
+			loop {
+				let updates_available = manager.wait_timeout(Duration::from_millis(100));
+				if updates_available {
+					if let Err(e) = persist_manager(manager.clone()) {
+						panic!("Errored persisting manager: {}", e);
+					};
+				}
+				// If we see that the thread has been stopped, exit now.
+				if stop_thread.load(Ordering::Relaxed) == true {
+					log_trace!(logger, "Terminating background processor.");
+					break;
+				}
+				if current_time.elapsed().as_secs() > CHAN_FRESHNESS_TIMER {
+					log_trace!(logger, "Calling manager's timer_chan_freshness_every_min");
+					manager.timer_chan_freshness_every_min();
```

Comment:
I know its (more than) a little awkward, but does it make sense to also have an argument that takes an `Option`al `PeerManager` and calls `timer_tick_occured` on it? It seems like a waste not to, but it also seems like a layer violation to. Maybe @jkczyz has a strong opinion?

Reply:
Seems we're billing this crate more as a general purpose background processor, so may make sense to add it. Is your concern around depending on waiting for `ChannelManager` at the start of the loop? Or that the persistence may delay calling `timer_tick_occured`. The latter may be a good reason why these should be separated.

Reply:
I started implementing it and does seem a tad awkward (doable though of course). Could I push this to a follow-up PR?

Reply:
My concern was it felt strange that this thing that basically just persists the channelmanager also needs a ref to a peerhandler. But the utility gain seems worth it. No issue with it being a follow-up.